### PR TITLE
send a batch of kv in encodeLoop

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve? 
屏幕快照 2020-03-10 上午10 41 34
About 58% cpu time is costed by runtime.selectgo, we can send a batch of kv pairs to optimize it.

### What is changed and how it works?
Send a batch of kv in encodeLoop instead of sending 1 kv pair every time.
After this commit, the cost of runtime.selectgo can be eliminated.
屏幕快照 2020-03-12 下午4 33 06

### Check List 

Tests 

exists tests
 
Side effects

N/A

Related changes

N/A

### CODE REVIEW

File: lightning/restore/restore_test.go [hunk 1]
```diff
@@ -742,30 +743,31 @@ func (s *chunkRestoreSuite) TestDeliverLoop(c *C) {
...
+	rc := &RestoreController{pauser: DeliverPauser, cfg: cfg}
+	_, _, err := s.cr.encodeLoop(ctx, kvsCh, s.tr, s.tr.logger, kvEncoder, deliverCompleteCh, rc)
 	c.Assert(err, IsNil)
 	c.Assert(kvsCh, HasLen, 2)
 
-	firstKVs := <-kvsCh
-	c.Assert(firstKVs.kvs, HasLen, 2)
-	c.Assert(firstKVs.rowID, Equals, int64(19))
-	c.Assert(firstKVs.offset, Equals, int64(36))
+	kvs := <-kvsCh
```
Comment (kennytm) [inline lightning/restore/restore_test.go]:
```suggestion
	kvs := <-kvsCh
	c.Assert(kvs, HasLen, 1)
```

**Comment by kennytm** [comment]
Test failed

```
[2020-03-10T21:47:50.192Z] Error: in file tests/default-columns/data/defcol.u.1.sql:0 at offset 30: failed to cast `<nil>` as int(11) for column `yy` (#2): [table:1364]Field 'yy' doesn't have a default value
```

**Comment by GregoryIan** [comment]
@XuHuaiyu please fix CI

**Comment by july2993** [comment]
> Test failed
> 
> ```
> [2020-03-10T21:47:50.192Z] Error: in file tests/default-columns/data/defcol.u.1.sql:0 at offset 30: failed to cast `<nil>` as int(11) for column `yy` (#2): [table:1364]Field 'yy' doesn't have a default value
> ```

fix by https://github.com/pingcap/tidb-lightning/pull/279/commits/26cfe9678a6f6a9a9dfca859f0a55a6a8cf2dbcf

pass wrong columnNames to `initializeColumns` -> get the wrong column permutation -> get the wrong column value

File: lightning/restore/restore.go [hunk 2]
```diff
@@ -1700,63 +1701,73 @@ func (cr *chunkRestore) encodeLoop(
...
+			lastRow := cr.parser.LastRow()
+			// sql -> kv
+			kvs, encodeErr := kvEncoder.Encode(logger, lastRow.Row, lastRow.RowID, cr.chunk.ColumnPermutation)
+			encodeDur += time.Since(encodeDurStart)
+			if encodeErr != nil {
+				err = errors.Annotatef(encodeErr, "in file %s at offset %d", &cr.chunk.Key, newOffset)
+				return
+			}
+			kvPacket = append(kvPacket, deliveredKVs{kvs: kvs, columns: columnNames, offset: newOffset, rowID: rowID})
+			if len(kvPacket) == int(maxKvPairsCnt) || newOffset == cr.chunk.Chunk.EndOffset {
```
Comment (kennytm) [inline lightning/restore/restore.go]:
```suggestion
			if len(kvPacket) >= maxKvPairsCnt || newOffset == cr.chunk.Chunk.EndOffset {
```

**Pull request merged**